File: mirofish/src/model/predictor.py
# src/model/predictor.py
import joblib
import numpy as np
from pathlib import Path
from src.features.schema import FeatureSet
from src.config import settings


class ChurnPredictor:
    def __init__(self):
        self.model = None
        self.feature_columns = settings.FEATURE_COLUMNS
        self._load_model()

    def _load_model(self):
        """Load the trained model from disk."""
        model_path = Path(settings.MODEL_PATH)
        if model_path.exists():
            self.model = joblib.load(model_path)
            print(f"Model loaded from {settings.MODEL_PATH}")
        else:
            print(f"Model not found at {settings.MODEL_PATH}. Using fallback logic.")
            self.model = None

    def predict_proba(self, features: FeatureSet) -> float:
        """Predict probability of churn (0-1)."""
        if self.model is None:
            # Fallback: rule-based prediction
            return self._fallback_prediction(features)

        # Prepare features array
        feature_values = [
            features.dias_ultima_visita,
            features.frequencia_mensal,
            features.ticket_medio,
            features.variabilidade_frequencia,
            features.tendencia_declinio,
            features.valor_total_comprado,
            features.idade_cadastro
        ]

        X = np.array([feature_values])
        proba = self.model.predict_proba(X)[0][1]  # Probability of class 1 (churn)
        return min(max(proba, 0.0), 1.0)  # Clamp to 0-1

    def _fallback_prediction(self, features: FeatureSet) -> float:
        """Simple rule-based prediction when model not available."""
        score = 0

        # Score by days since last visit
        if features.dias_ultima_visita > 90:
            score += 40
        elif features.dias_ultima_visita > 60:
            score += 25
        elif features.dias_ultima_visita > 30:
            score += 10

        # Score by frequency decline
        if features.tendencia_declinio < 0.5:
            score += 30
        elif features.tendencia_declinio < 0.8:
            score += 15

        # Score by low frequency
        if features.frequencia_mensal < 0.3:
            score += 20

        # Score by low ticket
        if features.ticket_medio < 50:
            score += 10

        return min(score / 100, 1.0)

    def get_risk_level(self, proba: float) -> tuple:
        """Get risk level and score from probability."""
        score = int(proba * 100)

        if score < 30:
            return ('baixo', score)
        elif score < 60:
            return ('medio', score)
        elif score < 80:
            return ('alto', score)
        else:
            return ('critico', score)

    def needs_retraining(self) -> bool:
        """Check if model file exists."""
        return not Path(settings.MODEL_PATH).exists()


# Global instance
predictor = ChurnPredictor()

File: src/lib/churn/predictor.ts
import { ChurnFeatures } from "./features"

/**
 * Risk levels based on score
 * 0-30: baixo (low risk)
 * 31-60: medio (medium risk)
 * 61-80: alto (high risk)
 * 81-100: critico (critical risk)
 */
export type ChurnRiskLevel = "baixo" | "medio" | "alto" | "critico"

/**
 * Churn status based on prediction
 */
export type ChurnStatus = "ativo" | "em_risco" | "em_risco_alto" | "perdido"

/**
 * Prediction result
 */
export type ChurnPredictionResult = {
  score: number // 0-100
  probability: number // 0-1
  level: ChurnRiskLevel
  status: ChurnStatus
}

/**
 * Simple rule-based churn predictor
 * When ML model is not available, uses rules to estimate churn risk
 */
export class ChurnPredictor {
  /**
   * Predict churn based on features
   * Uses rule-based approach when ML model is not available
   */
  predict(features: ChurnFeatures): ChurnPredictionResult {
    // Calculate base score from features
    let score = 0

    // Dias desde última visita (maior impacto)
    if (features.diasUltimaVisita > 90) {
      score += 40
    } else if (features.diasUltimaVisita > 60) {
      score += 25
    } else if (features.diasUltimaVisita > 30) {
      score += 10
    }

    // Tendência de declínio
    if (features.tendenciaDeclinio === 1) {
      score += 20
    }

    // Valor total comprado (clients que gastam pouco são mais propensos a churn)
    if (features.valorTotalComprado < 100) {
      score += 15
    } else if (features.valorTotalComprado < 500) {
      score += 5
    }

    // Idade do cadastro (clients novos podem ter maior churn)
    if (features.idadeCadastro < 30) {
      score += 10
    }

    // Frequência baixa
    if (features.frequenciaMensal < 1) {
      score += 15
    } else if (features.frequenciaMensal < 2) {
      score += 5
    }

    // Garantir score entre 0 e 100
    score = Math.min(100, Math.max(0, score))

    // Calculate probability (sigmoid-like transformation)
    const probability = 1 / (1 + Math.exp(-0.02 * (score - 50)))

    // Determine risk level
    let level: ChurnRiskLevel
    if (score <= 30) {
      level = "baixo"
    } else if (score <= 60) {
      level = "medio"
    } else if (score <= 80) {
      level = "alto"
    } else {
      level = "critico"
    }

    // Determine status
    let status: ChurnStatus
    if (score <= 30) {
      status = "ativo"
    } else if (score <= 60) {
      status = "em_risco"
    } else if (score <= 80) {
      status = "em_risco_alto"
    } else {
      status = "perdido"
    }

    return {
      score,
      probability,
      level,
      status,
    }
  }

  /**
   * Load ML model (placeholder for future implementation)
   */
  loadModel() {
    // Future implementation: Load XGBoost model from file
    // const model = await xgboost.Booster.load("data/models/churn_model.json")
    // return model
    return null
  }

  /**
   * Predict with ML model (placeholder for future implementation)
   */
  predictWithModel(features: ChurnFeatures): ChurnPredictionResult | null {
    // Future implementation: Use trained XGBoost model
    // const model = this.loadModel()
    // if (!model) return null
    // const dmatrix = xgboost.DMatrix([Object.values(features)])
    // const prediction = model.predict(dmatrix)
    // return {
    //   score: Math.round(prediction[0] * 100),
    //   probability: prediction[0],
    //   level: this.getRiskLevel(prediction[0]),
    //   status: this.getStatus(prediction[0]),
    // }
    return null
  }
}
